pkg/auth/providers/azure/azure_provider.go

```diff
@@ -354,20 +364,30 @@ func (ap *azureProvider) userGroupsToPrincipals(
 	azureClient *azureClient,
 	groups graphrbac.UserGetMemberGroupsResult,
 ) ([]v3.Principal, error) {
-	var groupPrincipals []v3.Principal
-	for _, group := range *groups.Value {
-		groupObj, err := azureClient.groupClient.Get(context.Background(), group)
-		if err != nil {
-			return groupPrincipals, err
-		}
+	start := time.Now()
+	logrus.Debug("[AZURE_PROVIDER] Started gathering users groups")
+	var g errgroup.Group
+	groupPrincipals := make([]v3.Principal, len(*groups.Value))
+	for i, group := range *groups.Value {
+		g.Go(func() error {
+			groupObj, err := azureClient.groupClient.Get(context.Background(), group)
+			if err != nil {
+				logrus.Debugf("[AZURE_PROVIDER] Error getting group: %v", err)
+				return err
+			}
 
-		p := ap.groupToPrincipal(groupObj)
-		p.MemberOf = true
+			p := ap.groupToPrincipal(groupObj)
+			p.MemberOf = true
 
-		groupPrincipals = append(groupPrincipals, p)
+			groupPrincipals[i] = p
```

Comment:
given `i` is a reference, there are chances it would hold a different value from what we expect it to have at this point, as the loop will keep on going